Shrink unicode case-mapping LUTs by 24k

martingms opened the pull request:
I was looking into the binary bloat of a small program using `str::to_lowercase` and `str::to_uppercase`, and noticed that the lookup tables used for case mapping had a lot of zero-bytes in them. The reason for this is that since some characters map to up to three other characters when lower or uppercased, the LUTs store a `[char; 3]` for each character. However, the vast majority of cases only map to a single new character, in other words most of the entries are e.g. `(lowerc, [upperc, '\0', '\0'])`.
This PR introduces a new encoding scheme for these tables.

The changes reduces the size of my test binary by about 24K.

I've also done some `#[bench]`marks on unicode-heavy test data, and found that the performance of both `str::to_lowercase` and `str::to_uppercase` improves by up to 20%. These measurements are obviously very dependent on the character distribution of the data.

Someone else will have to decide whether this more complex scheme is worth it or not, I was just goofing around a bit and here's what came out of it :man_shrugging: No hard feelings if this isn't wanted!

martingms committed:
Skip serializing ascii chars in case LUTs

Since ascii chars are already handled by a special case in the
`to_lower` and `to_upper` functions, there's no need to waste space on
them in the LUTs.

martingms committed:
Split unicode case LUTs in single and multi variants

The majority of char case replacements are single char replacements,
so storing them as [char; 3] wastes a lot of space.

This commit splits the replacement tables for both `to_lower` and
`to_upper` into two separate tables, one with single-character mappings
and one with multi-character mappings.

This reduces the binary size for programs using all of these tables
with roughly 24K bytes.

workingjubilee commented:
cc @thomcc

nikic commented:
It would be possible to avoid the second table lookup by storing an index into the table with the long expansions. Something like `(1 << 24) | large_index`. This does mean that it can't be stored as `char` and needs to use `u32` instead, because these would be stored in the invalid code point space.

martingms commented:
> It would be possible to avoid the second table lookup by storing an index into the table with the long expansions. Something like `(1 << 24) | large_index`. This does mean that it can't be stored as `char` and needs to use `u32` instead, because these would be stored in the invalid code point space.

That is a very good idea!

I'll see if I can get something working a bit later.

martingms committed:
Improve case mapping encoding scheme

The indices are encoded as `u32`s in the range of invalid `char`s, so
that we know that if any mapping fails to parse as a `char` we should
use the value for lookup in the multi-table.

This avoids the second binary search in cases where a multi-`char`
mapping is needed.

Idea from @nikic

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 355e1dda1df8a8106a1692eb7f019fb00df8c3ab with merge 7dff56ce084e1c8649af8ffed641f398d0e5ac9f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"355e1dda1df8a8106a1692eb7f019fb00df8c3ab","merge_sha":"7dff56ce084e1c8649af8ffed641f398d0e5ac9f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4456736268/jobs/7828243306)
Build commit: 7dff56ce084e1c8649af8ffed641f398d0e5ac9f (`7dff56ce084e1c8649af8ffed641f398d0e5ac9f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4456736268/jobs/7828243306"},"merge_sha":"7dff56ce084e1c8649af8ffed641f398d0e5ac9f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4456736268/jobs/7828243306)
Build commit: 7dff56ce084e1c8649af8ffed641f398d0e5ac9f (`7dff56ce084e1c8649af8ffed641f398d0e5ac9f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4456736268/jobs/7828243306"},"merge_sha":"7dff56ce084e1c8649af8ffed641f398d0e5ac9f"} -->

rust-timer commented:
Queued 7dff56ce084e1c8649af8ffed641f398d0e5ac9f with parent 2d0a7def330ed90f416a3429dbb44c5ee2a914e5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2d0a7def330ed90f416a3429dbb44c5ee2a914e5&end=7dff56ce084e1c8649af8ffed641f398d0e5ac9f).
<!-- rust-timer: temporary -->

Mark-Simulacrum commented:
r=me presuming perf looks ok

rust-timer commented:
Finished benchmarking commit (7dff56ce084e1c8649af8ffed641f398d0e5ac9f): [comparison URL](https://perf.rust-lang.org/compare.html?start=2d0a7def330ed90f416a3429dbb44c5ee2a914e5&end=7dff56ce084e1c8649af8ffed641f398d0e5ac9f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2d0a7def330ed90f416a3429dbb44c5ee2a914e5&end=7dff56ce084e1c8649af8ffed641f398d0e5ac9f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2d0a7def330ed90f416a3429dbb44c5ee2a914e5&end=7dff56ce084e1c8649af8ffed641f398d0e5ac9f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.2%  | [1.9%, 2.5%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.7%, -2.7%] | 3     |
| All ❌✅ (primary)                 | 0.8%  | [-2.1%, 2.5%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2d0a7def330ed90f416a3429dbb44c5ee2a914e5&end=7dff56ce084e1c8649af8ffed641f398d0e5ac9f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 355e1dda1df8a8106a1692eb7f019fb00df8c3ab has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 355e1dda1df8a8106a1692eb7f019fb00df8c3ab -->
<!-- homu: {"type":"Approved","sha":"355e1dda1df8a8106a1692eb7f019fb00df8c3ab","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/109163#issuecomment-1475007091) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

scottmcm commented on library/core/src/unicode/unicode_data.rs:
```diff
@@ -580,1815 +580,804 @@ pub mod white_space {
 
 #[rustfmt::skip]
 pub mod conversions {
+    const INDEX_MASK: u32 = 4194304;
```
nit: might be clearer if this were `0x40_0000` instead, at least for me who only knows my powers of two up so high.

martingms replied:
I agree, pushed a fix to `0x400000` at least (dunno if there's some fmt operator that inserts `_`s?).

Mark-Simulacrum replied:
Ultimately this is the generated code, so I'm not too worried about readability in it.

martingms committed:
Use hex literal for INDEX_MASK

Mark-Simulacrum commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 54f55efb9a147e8a7b5073d24c0cc67f0aad5a13 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 54f55efb9a147e8a7b5073d24c0cc67f0aad5a13 -->
<!-- homu: {"type":"Approved","sha":"54f55efb9a147e8a7b5073d24c0cc67f0aad5a13","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 54f55efb9a147e8a7b5073d24c0cc67f0aad5a13 with merge e5101b55d472f766f6e09e9077d59c575e31d3dd...
<!-- homu: {"type":"BuildStarted","head_sha":"54f55efb9a147e8a7b5073d24c0cc67f0aad5a13","merge_sha":"e5101b55d472f766f6e09e9077d59c575e31d3dd"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4508023345/jobs/7936961374)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4508023345/jobs/7936961374","builder_name":"checks-actions"} -->

martingms commented:
> broken_heart Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4508023345/jobs/7936961374)

Seems unrelated?

bors commented:
:hourglass: Testing commit 54f55efb9a147e8a7b5073d24c0cc67f0aad5a13 with merge f421586eed77de266a3f99ffa8a5687b7d2d893c...
<!-- homu: {"type":"BuildStarted","head_sha":"54f55efb9a147e8a7b5073d24c0cc67f0aad5a13","merge_sha":"f421586eed77de266a3f99ffa8a5687b7d2d893c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4510249502/jobs/7943773319)
Approved by: Mark-Simulacrum
Pushing f421586eed77de266a3f99ffa8a5687b7d2d893c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4510249502/jobs/7943773319"},"merge_sha":"f421586eed77de266a3f99ffa8a5687b7d2d893c"} -->

rust-timer commented:
Finished benchmarking commit (f421586eed77de266a3f99ffa8a5687b7d2d893c): [comparison URL](https://perf.rust-lang.org/compare.html?start=c763eceae349c1d827d9cfbf5df21ca40b21c861&end=f421586eed77de266a3f99ffa8a5687b7d2d893c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c763eceae349c1d827d9cfbf5df21ca40b21c861&end=f421586eed77de266a3f99ffa8a5687b7d2d893c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.0%, -1.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c763eceae349c1d827d9cfbf5df21ca40b21c861&end=f421586eed77de266a3f99ffa8a5687b7d2d893c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [0.5%, 1.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -3.3% | [-3.3%, -3.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.9% | [-2.1%, -0.5%] | 5     |
| All ❌✅ (primary)                 | -3.3% | [-3.3%, -3.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c763eceae349c1d827d9cfbf5df21ca40b21c861&end=f421586eed77de266a3f99ffa8a5687b7d2d893c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.9% | [-1.3%, -0.4%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
